sw/host/opentitanlib/src/transport/hyperdebug/mod.rs

```diff
@@ -0,0 +1,325 @@
+                        }
+                    }
+                }
+            }
+        }
+        // Send Ctrl-C, followed by the command, then newline.  This will discard any previous
+        // partial input, before executing our command.
+        port.write(format!("\u{3}{}\n", cmd).as_bytes())?;
+        let mut seen_echo = false;
+        let mut len: usize = 0;
+        loop {
+            match port.read(&mut buf[len..128]) {
+                Ok(rc) => {
+                    len += rc;
+                    'outer: loop {
+                        for i in 0..len {
+                            if buf[i] == 10 {
+                                // Newline
+                                let mut end = i;
+                                if end > 0 && buf[end - 1] == 13 {
+                                    end -= 1;
+                                }
+                                let line = std::str::from_utf8(&buf[0..end])?;
+                                if seen_echo {
+                                    callback(line);
+                                } else {
+                                    let minmatch = std::cmp::min(6, cmd.len());
+                                    if line.len() >= minmatch && line[line.len() - minmatch..]
+                                        == cmd[cmd.len() - minmatch..] {
+                                        seen_echo = true;
+                                    }
+                                }
+                                buf.rotate_left(i + 1);
+                                len -= i + 1;
+                                continue 'outer;
+                            }
+                        }
+                        break
+                    }
+                }
+                Err(error) => {
+                    match error.kind() {
+                        std::io::ErrorKind::TimedOut => {
+                            if std::str::from_utf8(&buf[0..len])? == "> " {
+                                // No data arrived for a while, and the last we got was a
+                                // command prompt, this is what we expect when the command has
+                                // finished successfully.
+                                return Ok(())
+                            } else {
+                                return Err(error.into())
+                            }
+                        }
+                        _ => {
+                            return Err(error.into())
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("USB device did not match")]
+    NoMatch,
+    #[error("Found no HyperDebug USB device")]
+    NoDevice,
+    #[error("Found multiple HyperDebug USB devices, use --serial")]
+    MultipleDevices,
+    #[error("Error communicating with HyperDebug")]
+    CommunicationError,
```

Comment:
This particular error is used in many different places and mostly as a general error in dealing with initial setup of communications.  You might want to define this as:

```
  #[error("Error communicating with HyperDebug: {0}")]
  CommunicationError(String),
```

And then add a short string at each error site describing the nature of the error.  This will make debug easier in the event that something goes wrong with the device.

Reply:
I have done so. All the strings being literals, I went for &'static str, instead of String.